File: 0039_combination_sum/combination_sum.cpp
#include <iostream>
#include <algorithm>
#include <vector>

using namespace std;

void dfs(vector<int> &nums, vector<int> &sub, int index, int target,
         vector<vector<int>> &res)
{
    if (target == 0)
    {
        res.push_back(sub);
        return;
    }
    for (int i = index; i < nums.size(); i++)
    {
        if (nums[i] > target)
            break;
        sub.push_back(nums[i]);
        dfs(nums, sub, i, target - nums[i], res);
        sub.pop_back();
    }
}

/**
 * Time: O(n * log(n) + 2**n)
 * Space: O(n**2)
 */
vector<vector<int>> combination_sum(vector<int> &candidates, int target)
{
    vector<vector<int>> res;
    vector<int> sub;
    sort(candidates.begin(), candidates.end());
    dfs(candidates, sub, 0, target, res);
    return res;
}

File: 0040_combination_sum_II/combination_sum.py
from typing import List


def combination_sum(candidates: List[int], target: int) -> List[List[int]]:
    """
    Time:    O(n * log(n) + 2**n)
    Space:   O(n**2)
    """
    res = []
    nums = sorted(candidates)
    dfs(nums, 0, [], target, res)
    return res


def dfs(nums: List[int], index: int, arr: List[int], target: int,
        res: List[List[int]]):
    if not target:
        res.append(arr)
        return
    for i in range(index, len(nums)):
        curr = nums[i]
        if i > index and curr == nums[i - 1]:
            continue
        if curr > target:
            break
        dfs(nums, i + 1, arr + [curr], target - curr, res)
